### PR TITLE
Allow Go Dynamic Instrumentation to run on amd64

### PR DESCRIPTION
**Pull request by grantseltzer**
### What does this PR do?

- Enables the dynamic instrumentation system-probe module to on amd64 in addition to arm64.
    - For now parameters aren't being captured, they will once we correctly parse the ABI.
- Generalizes register/stack reads in bpf code to use custom defined macros
- Makes it so capturing of strings/slices uses specific registers for lengths instead of inferring it (by adding 1 to the register number or 8 to the stack offset)

### Motivation

Users often don't have control over what architecture the node they're deployed on.

### Describe how to test/QA your changes

Run recently merged KMT tests

### CODE REVIEW

**Comment by grantseltzer** [comment]
/trigger-ci --variable RUN_ALL_BUILDS=true --variable RUN_KMT_TESTS=on

File: pkg/dynamicinstrumentation/diconfig/binary_inspection.go [hunk 1]
```diff
@@ -73,6 +73,10 @@ func AnalyzeBinary(procInfo *ditypes.ProcessInfo) error {
 	if err != nil {
 		return fmt.Errorf("could not determine locations of variables from debug information %w", err)
 	}
+	stringPtrIdentifier := bininspect.FieldIdentifier{StructName: "string", FieldName: "str"}
+	stringLenIdentifier := bininspect.FieldIdentifier{StructName: "string", FieldName: "len"}
+	r.StructOffsets[stringPtrIdentifier] = 0
+	r.StructOffsets[stringLenIdentifier] = 8
```
Comment (cimi) [inline pkg/dynamicinstrumentation/diconfig/binary_inspection.go:79]: Why do we need to set these, won't `InspectWithDWARF` get them?

File: pkg/dynamicinstrumentation/diconfig/binary_inspection.go [hunk 2]
```diff
@@ -203,7 +226,7 @@ func correctTypeSpecificLocations(params []ditypes.Parameter, fieldLocations map
 			correctArrayLocations(&params[i], fieldLocations)
 		} else if params[i].Kind == uint(reflect.Pointer) {
 			correctPointerLocations(&params[i], fieldLocations)
-		} else if params[i].Kind == uint(reflect.Struct) {
+		} else if params[i].Kind == uint(reflect.Struct) || params[i].Kind == uint(reflect.String) {
 			correctStructLocations(&params[i], fieldLocations)
 		}
```
Comment (cimi) [inline pkg/dynamicinstrumentation/diconfig/binary_inspection.go:231]: Do we need to special case slices here?

File: pkg/dynamicinstrumentation/codegen/codegen.go [hunk 3]
```diff
@@ -223,10 +239,81 @@ func generateSliceHeader(slice *ditypes.Parameter, out io.Writer) error {
...
+	buf := bytes.NewBuffer(x)
+	err := generateStringLengthHeader(stringParam.ParameterPieces[1], buf)
+	if err != nil {
+		return err
+	}
+
+	stringHeaderWrapper := stringHeaderWrapper{
+		Parameter:        stringParam,
+		StringLengthText: buf.String(),
+	}
```
Comment (cimi) [inline pkg/dynamicinstrumentation/codegen/codegen.go:264]: What do we gain by separating the length template from the main header one and injecting the length as text?

**Review by cimi — COMMENTED** [review]
:+1: Looks good, left a few comments

Since we're not capturing params on amd64 yet, it would have been better to have the string/slice changes in a separate PR.

File: pkg/ebpf/c/bpf_tracing_custom.h [hunk 4]
```diff
@@ -26,6 +46,41 @@
     p;                                                                     \
 })
 
+#define DI_REGISTER_0 regs[0]
```
Comment (usamasaqib) [inline pkg/ebpf/c/bpf_tracing_custom.h:49]: Can we define them with the function syntax?
```
#define DI_REGISTER(num) regs[num]
```

File: pkg/ebpf/c/bpf_tracing_custom.h [hunk 5]
```diff
@@ -16,6 +16,26 @@
 #define PT_REGS_PARM9(x) PT_REGS_STACK_PARM(x,3)
 #define PT_REGS_PARM10(x) PT_REGS_STACK_PARM(x,4)
 
+#define DI_REGISTER_0 ax
```
Comment (usamasaqib) [inline pkg/ebpf/c/bpf_tracing_custom.h:19]: Is there any specific reason behind the ordering?
For me, I would expect the registers to be ordered by the architecture's calling convention, for example.

Comment (grantseltzer) [reply pkg/ebpf/c/bpf_tracing_custom.h:19]: Yea it is the x86 ABI: https://refspecs.linuxfoundation.org/elf/x86_64-abi-0.95.pdf#page=56

File: pkg/dynamicinstrumentation/diconfig/config_manager.go [hunk 6]
```diff
@@ -249,6 +249,12 @@ generateCompileAttach:
 	err = codegen.GenerateBPFParamsCode(procInfo, probe)
 	if err != nil {
 		log.Info("Couldn't generate BPF programs", err)
+		if !probe.InstrumentationInfo.AttemptedRebuild {
+			log.Info("Removing parameters and attempting to rebuild BPF object", err)
+			probe.InstrumentationInfo.AttemptedRebuild = true
+			probe.InstrumentationInfo.InstrumentationOptions.CaptureParameters = false
+			goto generateCompileAttach
+		}
```
[the remark this answers is not in the document]
Comment (grantseltzer) [reply pkg/dynamicinstrumentation/diconfig/config_manager.go:257]: There were errors occurring while I was making some refactoring changes that shouldn't happen any more, but still, code generation can return an error, and if it does we should have a way of backing off.

File: pkg/dynamicinstrumentation/diconfig/binary_inspection.go (hunk 1 above)
Comment (grantseltzer) [reply pkg/dynamicinstrumentation/diconfig/binary_inspection.go:79]: I need to investigate further, but InspectWithDWARF was not getting them.

File: pkg/dynamicinstrumentation/codegen/codegen.go (hunk 3 above)
Comment (grantseltzer) [reply pkg/dynamicinstrumentation/codegen/codegen.go:259]: Precisely reading the length directly from where it's passed. What we were doing before was looking at what register the `ptr` was passed in, and adding 1 to it. This works on ARM but not x86 (the x86 ABI is strange).

File: pkg/dynamicinstrumentation/diconfig/binary_inspection_test.go [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+		t.Error(">", err)
+	}
+
+	for _, funcMetadata := range result.Functions {
+		for paramName, paramMeta := range funcMetadata.Parameters {
+			for _, piece := range paramMeta.Pieces {
+				pretty.Log(paramName, piece)
+			}
+		}
+	}
```
[the remark this answers is not in the document]
Comment (grantseltzer) [reply pkg/dynamicinstrumentation/diconfig/binary_inspection_test.go:59]: Yes - I was just using this for development but i'll turn it into an actual test.

File: pkg/dynamicinstrumentation/uploader/di_log_converter.go [hunk 8]
```diff
@@ -121,14 +134,26 @@ func convertArgs(defs []ditypes.Parameter, captures []*ditypes.Param) map[string
 			valueCopy := capture.ValueStr
 			cv.Value = &valueCopy
 		}
-		if capture.Fields != nil && idx < len(defs) {
+		if capture.Fields != nil {
```
[the remark this answers is not in the document]
Comment (grantseltzer) [reply pkg/dynamicinstrumentation/uploader/di_log_converter.go:133]: Good call, I believe this is just leftover from debugging/development but will fix it.

File: pkg/ebpf/c/bpf_tracing_custom.h (hunk 5 above)
Comment (brycekahle) [reply pkg/ebpf/c/bpf_tracing_custom.h:19]: It is the mapping between DWARF register numbers and architecture register names. Can you add the link you posted as a comment?

Comment (brycekahle) [reply pkg/ebpf/c/bpf_tracing_custom.h:19]: maybe naming them `DWARF_REGISTER_X` would make it clearer?

File: pkg/dynamicinstrumentation/codegen/c/dynamicinstrumentation.c [hunk 9]
```diff
@@ -64,12 +64,12 @@ int {{.GetBPFFuncName}}(struct pt_regs *ctx)
     event->base.uid = uid;
 
     // Collect stack trace
-    __u64 currentPC = ctx->pc;
+    __u64 currentPC = ctx->__PT_IP_REG;
```
Comment (brycekahle) [inline pkg/dynamicinstrumentation/codegen/c/dynamicinstrumentation.c:67]: I think the preferred way is this: 
```suggestion
    __u64 currentPC = PT_REGS_IP(ctx);
```

File: pkg/dynamicinstrumentation/codegen/c/dynamicinstrumentation.c [hunk 10]
```diff
@@ -64,12 +64,12 @@ int {{.GetBPFFuncName}}(struct pt_regs *ctx)
...
     bpf_probe_read(&event->base.program_counters[0], sizeof(__u64), &currentPC);
 
-    __u64 bp = ctx->regs[29];
+    __u64 bp = ctx->__PT_FP_REG;
```
Comment (brycekahle) [inline pkg/dynamicinstrumentation/codegen/c/dynamicinstrumentation.c:70]:
```suggestion
    __u64 bp = PT_REGS_FP(ctx);
```

File: pkg/dynamicinstrumentation/codegen/c/dynamicinstrumentation.c [hunk 11]
```diff
@@ -64,12 +64,12 @@ int {{.GetBPFFuncName}}(struct pt_regs *ctx)
...
     bpf_probe_read(&bp, sizeof(__u64), (void*)bp); // dereference bp to get current stack frame
-    __u64 ret_addr = ctx->regs[30]; // when bpf prog enters, the return address hasn't yet been written to the stack
+    __u64 ret_addr = ctx->__PT_RET_REG; // when bpf prog enters, the return address hasn't yet been written to the stack
```
Comment (brycekahle) [inline pkg/dynamicinstrumentation/codegen/c/dynamicinstrumentation.c:72]:
```suggestion
    __u64 ret_addr = PT_REGS_RET(ctx); // when bpf prog enters, the return address hasn't yet been written to the stack
```

**Review by brycekahle — APPROVED** [review]

**Comment by grantseltzer** [comment]
@usamasaqib @cimi I think I need your approvals to merge?

**Review by cimi — APPROVED** [review]

**Pull request merged**